# Tensorflow-2.0 in Action 

![title](img/tensorflow1.png)

> TensorFlow is an open source software library for high performance numerical computation. Its flexible architecture allows easy deployment of computation across a variety of platforms (CPUs, GPUs, TPUs), and from desktops to clusters of servers to mobile and edge devices.
> Originally developed by researchers and engineers from the Google Brain team within Google’s AI organization, it comes with strong support for machine learning and deep learning and the flexible numerical computation core is used across many other scientific domains.

## Why Tensorflow?

![title](img/comparision.png)

![title](img/comparision2.png)

## Installation of Tensorflow

>TensorFlow is tested and supported on the following 64-bit systems:

>1.Ubuntu 16.04 or later

>2.Windows 7 or later

>3.macOS 10.12.6 (Sierra) or later (no GPU support)

>4.Raspbian 9.0 or later

> **pip install tensorflow==2.0.0**

> To run from Anaconda Prompt

> **!pip install tensorflow==2.0.0**

> To run from Jupyter Notebook

`Both Tensorflow 2.0 and Keras have been released for four years (Keras was released in March 2015, and Tensorflow was released in November of the same year). The rapid development of deep learning in the past days, we also know some problems of Tensorflow1.x and Keras:`

* Using Tensorflow means programming static graphs, which is difficult and inconvenient for programs that are familiar with imperative programming

* Tensorflow api is powerful and flexible, but it is more complex, confusing and difficult to use.

* Keras api is productive and easy to use, but lacks flexibility for research


#### Version Check

In [1]:
import tensorflow as tf
print("TensorFlow version: {}".format(tf.__version__))
print("Keras version: {}".format(tf.keras.__version__))

TensorFlow version: 2.11.0
Keras version: 2.11.0


`Tensorflow2.0 is a combination design of Tensorflow1.x and Keras. Considering user feedback and framework development over the past four years, it largely solves the above problems and will become the future machine learning platform.`

> Tensorflow 2.0 is built on the following core ideas:


* The coding is more pythonic, so that users can get the results immediately like they are programming in numpy
* Retaining the characteristics of static graphs (for performance, distributed, and production deployment), this makes TensorFlow fast, scalable, and ready for production.
* Using Keras as a high-level API for deep learning, making Tensorflow easy to use and efficient
* Make the entire framework both high-level features (easy to use, efficient, and not flexible) and low-level features (powerful and scalable, not easy to use, but very flexible)

>Eager execution is by default in TensorFlow 2.0 and, it needs no special setup.
>The following below code can be used to find out whether a CPU or GPU is in use

### GPU/CPU Check

In [2]:
variable = tf.Variable([3, 3])
if tf.test.is_gpu_available():
    print('GPU')
    print('GPU #0?')
    print(variable.device.endswith('GPU:0'))
else:
    print('CPU')

Instructions for updating:
Use `tf.config.list_physical_devices('GPU')` instead.


GPU
GPU #0?
False


### Tensor Constant

In [3]:
variable

<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([3, 3], dtype=int32)>

In [4]:
type(variable)

tensorflow.python.ops.resource_variable_ops.ResourceVariable

In [5]:
ineuron = tf.constant(42)
type(ineuron)

tensorflow.python.framework.ops.EagerTensor

In [6]:
ineuron.numpy()

42

In [7]:
ineuron.dtype

tf.int32

In [8]:
ineuron.shape

TensorShape([])

In [9]:
ineuron1 = tf.constant(1, dtype = tf.int64)
ineuron1

<tf.Tensor: shape=(), dtype=int64, numpy=1>

In [10]:
ineuron_x = tf.constant([[4,2],[9,5]])
print(ineuron_x)

tf.Tensor(
[[4 2]
 [9 5]], shape=(2, 2), dtype=int32)


In [11]:
ineuron_x.numpy()

array([[4, 2],
       [9, 5]], dtype=int32)

In [12]:
print('shape:',ineuron_x.shape)
print(ineuron_x.dtype)

shape: (2, 2)
<dtype: 'int32'>


#### Commonly used method is to generate constant tf.ones and the tf.zeros like of numpy np.ones & np.zeros

In [13]:
print(tf.ones(shape=(2,3)))

tf.Tensor(
[[1. 1. 1.]
 [1. 1. 1.]], shape=(2, 3), dtype=float32)


In [14]:
print(tf.zeros(shape=(3,2)))

tf.Tensor(
[[0. 0.]
 [0. 0.]
 [0. 0.]], shape=(3, 2), dtype=float32)


In [15]:
import tensorflow as tf
const2 = tf.constant([[3,4,5], [3,4,5]]);tf
const1 = tf.constant([[1,2,3], [1,2,3]]);
result = tf.add(const1, const2);
print(result)

tf.Tensor(
[[4 6 8]
 [4 6 8]], shape=(2, 3), dtype=int32)


>We have defined two constants and we add one value to the other. 
>As a result, we got a Tensor object with the result of the adding. 

#### Random constant

In [16]:
tf.random.normal(shape=(2,2),mean=0,stddev=1.0)

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[ 0.11143883, -1.2921705 ],
       [-0.15593928,  0.06588898]], dtype=float32)>

In [17]:
tf.random.uniform(shape=(2,2),minval=0,maxval=10,dtype=tf.int32)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[9, 8],
       [2, 5]], dtype=int32)>

### Variables

>A variable is a special tensor that is used to store variable values ​​and needs to be initialized with some values

#### Declaring variables

In [18]:
var0 = 24 # python variable
var1 = tf.Variable(42) # rank 0 tensor
var2 = tf.Variable([ [ [0., 1., 2.], [3., 4., 5.] ], [ [6., 7., 8.], [9., 10., 11.] ] ]) #rank 3 tensor
var0, var1, var2

(24,
 <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=42>,
 <tf.Variable 'Variable:0' shape=(2, 2, 3) dtype=float32, numpy=
 array([[[ 0.,  1.,  2.],
         [ 3.,  4.,  5.]],
 
        [[ 6.,  7.,  8.],
         [ 9., 10., 11.]]], dtype=float32)>)

>TensorFlow will infer the datatype, defaulting to tf.float32 for floats and tf.int32 for integers

#### The datatype can be explicitly specified

In [19]:
float_var64 = tf.Variable(89, dtype = tf.float64)
float_var64.dtype

tf.float64

`TensorFlow has a large number of built-in datatypes.`
* tf.float16: 16-bit half-precision floating-point.
* tf.float32: 32-bit single-precision floating-point.
* tf.float64: 64-bit double-precision floating-point.
* tf.bfloat16: 16-bit truncated floating-point.
* tf.complex64: 64-bit single-precision complex.
* tf.complex128: 128-bit double-precision complex.
* tf.int8: 8-bit signed integer.
* tf.uint8: 8-bit unsigned integer.
* tf.uint16: 16-bit unsigned integer.
* tf.uint32: 32-bit unsigned integer.
* tf.uint64: 64-bit unsigned integer.
* tf.int16: 16-bit signed integer.
* tf.int32: 32-bit signed integer.
* tf.int64: 64-bit signed integer.
* tf.bool: Boolean.
* tf.string: String.
* tf.qint8: Quantized 8-bit signed integer.
* tf.quint8: Quantized 8-bit unsigned integer.
* tf.qint16: Quantized 16-bit signed integer.
* tf.quint16: Quantized 16-bit unsigned integer.
* tf.qint32: Quantized 32-bit signed integer.
* tf.resource: Handle to a mutable resource.
* tf.variant: Values of arbitrary types.


#### To reassign a variable, use var.assign()

In [20]:
var_reassign = tf.Variable(89.)
var_reassign

<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=89.0>

In [21]:
type(var_reassign)

tensorflow.python.ops.resource_variable_ops.ResourceVariable

In [22]:
var_reassign.assign(98.)
var_reassign

<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=98.0>

In [23]:
initial_value = tf.random.normal(shape=(2,2))
a = tf.Variable(initial_value)
print(a)

<tf.Variable 'Variable:0' shape=(2, 2) dtype=float32, numpy=
array([[ 0.20591408, -1.6965824 ],
       [-1.318267  , -0.17015561]], dtype=float32)>


>We can assign "=" with assign (value), or assign_add (value) with "+ =", or assign_sub (value) with "-="

In [24]:
new_value = tf.random.normal(shape=(2, 2))
a.assign(new_value)
for i in range(2):
    for j in range(2):
        assert a[i, j] == new_value[i, j]

In [25]:
added_value = tf.random.normal(shape=(2,2))
a.assign_add(added_value)
for i in range(2):
    for j in range(2):
        assert a[i,j] == new_value[i,j]+added_value[i,j]

#### Shaping a tensor

In [26]:
tensor = tf.Variable([ [ [10., 11., 12.], [13., 14., 15.] ], [ [16., 17., 18.], [19., 20., 21.] ] ]) # tensor variable
print(tensor.shape)

(2, 2, 3)


#### Tensors can be reshaped and retain the same values which is required for constructing Neural networks.

In [27]:
tensor1 = tf.reshape(tensor,[2,6]) # 2 rows 6 cols
#tensor2 = tf.reshape(tensor,[1,12]) # 1 rows 12 cols
tensor1

<tf.Tensor: shape=(2, 6), dtype=float32, numpy=
array([[10., 11., 12., 13., 14., 15.],
       [16., 17., 18., 19., 20., 21.]], dtype=float32)>

In [28]:
print(tensor1.shape)

(2, 6)


In [29]:
tensor2 = tf.reshape(tensor,[1,12]) # 1 row 12 columns
tensor2

<tf.Tensor: shape=(1, 12), dtype=float32, numpy=
array([[10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20., 21.]],
      dtype=float32)>

### Rank of a tensor

>The rank of a tensor is defined as the number of dimensions, which is the number of indices that are required to specify any particular element of that tensor.

In [30]:
tensor

<tf.Variable 'Variable:0' shape=(2, 2, 3) dtype=float32, numpy=
array([[[10., 11., 12.],
        [13., 14., 15.]],

       [[16., 17., 18.],
        [19., 20., 21.]]], dtype=float32)>

In [31]:
tf.rank(tensor)

<tf.Tensor: shape=(), dtype=int32, numpy=3>

>(the shape is () because the output here is a scalar value)

#### Specifying an element of a tensor

In [32]:
tensor3 = tensor[1, 0, 2] # slice 1, row 0, column 2
tensor3

<tf.Tensor: shape=(), dtype=float32, numpy=18.0>

#### Casting a tensor to a NumPy variable

In [33]:
print(tensor.numpy())

[[[10. 11. 12.]
  [13. 14. 15.]]

 [[16. 17. 18.]
  [19. 20. 21.]]]


In [34]:
print(type(tensor.numpy()))

<class 'numpy.ndarray'>


In [35]:
print(tensor[1, 0, 2].numpy())

18.0


#### Finding the size or length of a tensor

In [36]:
tensor_size = tf.size(input=tensor).numpy()
tensor_size

12

In [37]:
#the datatype of a tensor
tensor3.dtype

tf.float32

### Tensorflow mathematical operations
>Can be used as numpy for artificial operations. Tensorflow can not execute these operations on the GPU or TPU.


In [38]:
a = tf.random.normal(shape=(2,2))
b = tf.random.normal(shape=(2,2))
c = a + b
d = tf.square(c)
e = tf.exp(c)
print(a)
print(b)
print(c)
print(d)
print(e)

tf.Tensor(
[[-0.5189267  -0.9635453 ]
 [-0.04750615 -0.97197026]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[ 0.0606637  -0.71593624]
 [-0.4525994  -1.3845032 ]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[-0.45826298 -1.6794815 ]
 [-0.50010556 -2.3564734 ]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[0.21000496 2.8206582 ]
 [0.25010556 5.552967  ]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[0.63238114 0.18647063]
 [0.60646665 0.09475379]], shape=(2, 2), dtype=float32)


### Performing element-wise primitive tensor operations

In [39]:
tensor

<tf.Variable 'Variable:0' shape=(2, 2, 3) dtype=float32, numpy=
array([[[10., 11., 12.],
        [13., 14., 15.]],

       [[16., 17., 18.],
        [19., 20., 21.]]], dtype=float32)>

In [40]:
tensor*tensor

<tf.Tensor: shape=(2, 2, 3), dtype=float32, numpy=
array([[[100., 121., 144.],
        [169., 196., 225.]],

       [[256., 289., 324.],
        [361., 400., 441.]]], dtype=float32)>

### Broadcasting in Tensorflow

>Element-wise tensor operations support broadcasting in the same way that NumPy arrays do.

>The simplest example is multiplication of  a tensor by a scalar value.

In [41]:
tensor4 = tensor*4
print(tensor4)

tf.Tensor(
[[[40. 44. 48.]
  [52. 56. 60.]]

 [[64. 68. 72.]
  [76. 80. 84.]]], shape=(2, 2, 3), dtype=float32)


### Transpose Matrix multiplication

In [42]:
matrix_u = tf.constant([[6,7,6]])
matrix_v = tf.constant([[3,4,3]])
tf.matmul(matrix_u, tf.transpose(a=matrix_v))

<tf.Tensor: shape=(1, 1), dtype=int32, numpy=array([[64]], dtype=int32)>

### Casting a tensor to another datatype

In [43]:
tensor1

<tf.Tensor: shape=(2, 6), dtype=float32, numpy=
array([[10., 11., 12., 13., 14., 15.],
       [16., 17., 18., 19., 20., 21.]], dtype=float32)>

In [44]:
i = tf.cast(tensor1, dtype=tf.int32)
i

<tf.Tensor: shape=(2, 6), dtype=int32, numpy=
array([[10, 11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20, 21]], dtype=int32)>

##### Casting with truncation

In [45]:
j = tf.cast(tf.constant(5.9), dtype=tf.int32)
j

<tf.Tensor: shape=(), dtype=int32, numpy=5>

###  Ragged tensors

`A ragged tensor is a tensor having one or more ragged dimensions. Ragged dimensions are dimensions that have slices having various lengths.There are a variety of methods for the declaration of ragged arrays, the simplest way is declaring a constant ragged array.`

#### Below example shows how to declare a constant ragged array

In [46]:
ragged =tf.ragged.constant([[9, 7, 4, 3], [], [11, 12, 8], [3], [7,8]])
print(ragged)
print(ragged[0,:])
print(ragged[1,:])
print(ragged[2,:])
print(ragged[3,:])
print(ragged[4,:])

<tf.RaggedTensor [[9, 7, 4, 3], [], [11, 12, 8], [3], [7, 8]]>
tf.Tensor([9 7 4 3], shape=(4,), dtype=int32)
tf.Tensor([], shape=(0,), dtype=int32)
tf.Tensor([11 12  8], shape=(3,), dtype=int32)
tf.Tensor([3], shape=(1,), dtype=int32)
tf.Tensor([7 8], shape=(2,), dtype=int32)


### Squared difference of tensors

In [47]:
varx = [4,5,6,1,2]
vary = 8
varz = tf.math.squared_difference(varx,vary)
varz

<tf.Tensor: shape=(5,), dtype=int32, numpy=array([16,  9,  4, 49, 36], dtype=int32)>

#### Calculate the mean

>Function available
>tf.reduce_mean().

`Similar to np.mean, except that it infers the return datatype from the input tensor, whereas np.mean allows you to specify the output type`

`tf.reduce_mean(input_tensor, axis=None, keepdims=None, name=None)`

In [48]:
# Defining a constant
numbers = tf.constant([[8., 9.], [1., 2.]])
numbers

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[8., 9.],
       [1., 2.]], dtype=float32)>

#### Calculate the mean across all axes

In [49]:
tf.reduce_mean(input_tensor=numbers) #default axis = None

<tf.Tensor: shape=(), dtype=float32, numpy=5.0>

#### Calculate the mean across columns (reduce rows) with this:

In [50]:
tf.reduce_mean(input_tensor=numbers, axis=0)

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([4.5, 5.5], dtype=float32)>

#### When keepdims = True

In [51]:
tf.reduce_mean(input_tensor=numbers, axis=0, keepdims=True) #the reduced axis is retained with a length of 1

<tf.Tensor: shape=(1, 2), dtype=float32, numpy=array([[4.5, 5.5]], dtype=float32)>

#### Calculate the mean across rows (reduce columns) with this:

In [52]:
tf.reduce_mean(input_tensor=numbers, axis=1)

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([8.5, 1.5], dtype=float32)>

#### When keepdims= True

In [53]:
tf.reduce_mean(input_tensor=numbers, axis=1, keepdims=True) #the reduced axis is retained with a length of 1

<tf.Tensor: shape=(2, 1), dtype=float32, numpy=
array([[8.5],
       [1.5]], dtype=float32)>

####  Random values generation

##### tf.random.normal()

>tf.random.normal() outputs a tensor of the given shape filled with values of the dtype type from a normal distribution.

>The function is as follows:
    
`tf.random.normal(shape, mean = 0, stddev =2, dtype=tf.float32, seed=None, name=None)`

In [54]:
tf.random.normal(shape = (3,2), mean=10, stddev=2, dtype=tf.float32, seed=None, name=None)
randon_num = tf.random.normal(shape = (3,2), mean=10.0, stddev=2.0)
print(randon_num)

tf.Tensor(
[[ 8.193186  9.814586]
 [12.509546 10.656453]
 [ 9.270586 13.391436]], shape=(3, 2), dtype=float32)


####  tf.random.uniform()

>The function is this:
    
>tf.random.uniform(shape, minval = 0, maxval= None, dtype=tf.float32, seed=None, name=None)

`This outputs a tensor of the given shape filled with values from a uniform distribution in the range minval to maxval, where the lower bound is inclusive but the upper bound isn't.

Example:`

In [55]:
tf.random.uniform(shape = (2,4), minval=0, maxval=None, dtype=tf.float32, seed=None, name=None)

<tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[0.33635283, 0.7145159 , 0.81271386, 0.43794334],
       [0.04875088, 0.4154415 , 0.848294  , 0.3638115 ]], dtype=float32)>

#### Setting the seed

In [56]:
tf.random.set_seed(12)
random_num1 = tf.random.uniform(shape = (2,2), maxval=10, dtype = tf.int32)
random_num2 = tf.random.uniform(shape = (2,2), maxval=10, dtype = tf.int32)
print(random_num1) #Call 1
print(random_num2)

tf.Tensor(
[[2 2]
 [6 2]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[8 2]
 [2 2]], shape=(2, 2), dtype=int32)


In [57]:
tf.random.set_seed(11) #same seed
random_num1 = tf.random.uniform(shape = (2,2), maxval=10, dtype = tf.int32)
random_num2 = tf.random.uniform(shape = (2,2), maxval=10, dtype = tf.int32)
print(random_num1) #Call 2
print(random_num2)

tf.Tensor(
[[4 6]
 [5 2]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[9 7]
 [9 4]], shape=(2, 2), dtype=int32)


#### Practical example of Random values using Dices

In [58]:
dice11 = tf.Variable(tf.random.uniform([10, 1], minval=1, maxval=7, dtype=tf.int32))
dice12 = tf.Variable(tf.random.uniform([10, 1], minval=1, maxval=7, dtype=tf.int32))

# lets ADD
dice_sum1 = dice11 + dice12
# We've got three separate 10x1 matrices. To produce a single 10x3 matrix, we'll concatenate them along dimension 1.
finale_matrix = tf.concat(values=[dice11, dice12, dice_sum1], axis=1)
print(finale_matrix)

tf.Tensor(
[[ 5  5 10]
 [ 4  3  7]
 [ 5  3  8]
 [ 3  3  6]
 [ 1  4  5]
 [ 4  1  5]
 [ 5  1  6]
 [ 6  4 10]
 [ 3  3  6]
 [ 2  3  5]], shape=(10, 3), dtype=int32)


#### Finding the indices of the largest and smallest element

>The following functions are available:
    
>`tf.argmax(input, axis=None, name=None, output_type=tf.int64 )`

>`tf.argmin(input, axis=None, name=None, output_type=tf.int64 )`

In [59]:
# 1-D tensor
tensor_1d = tf.constant([12, 11, 51, 42, 6, 16, -8, -19, 31])
print(tensor_1d)


i = tf.argmax(input=tensor_1d)
print('index of max; ', i)
print('Max element: ',tensor_1d[i].numpy())


i = tf.argmin(input=tensor_1d,axis=0).numpy()
print('index of min: ', i)
print('Min element: ',tensor_1d[i].numpy())

tf.Tensor([ 12  11  51  42   6  16  -8 -19  31], shape=(9,), dtype=int32)
index of max;  tf.Tensor(2, shape=(), dtype=int64)
Max element:  51
index of min:  7
Min element:  -19


#### Saving and restoring using a checkpoint

In [60]:
variable1 = tf.Variable([[5,6,9,3],[14,15,16,18]])
checkpoint= tf.train.Checkpoint(var=variable1)
savepath = checkpoint.save('./vars')
variable1.assign([[0,0,0,0],[0,0,0,0]])
variable1
checkpoint.restore(savepath)
print(variable1)

<tf.Variable 'Variable:0' shape=(2, 4) dtype=int32, numpy=
array([[ 5,  6,  9,  3],
       [14, 15, 16, 18]], dtype=int32)>


#### Using tf.function

`tf.function is a function that will take a Python function and return a TensorFlow graph. The advantage of this is that graphs can apply optimizations and exploit parallelism in the Python function (func). tf.function is new to TensorFlow 2.`


>Its function is as follows:
    
`tf.function(
func=None,input_signature=None,autograph=True,experimental_autograph_options=None
)`


In [61]:
def f1(x, y):
    return tf.reduce_mean(input_tensor=tf.multiply(x ** 3, 6) + y**3)

In [62]:
f1(4,5)

<tf.Tensor: shape=(), dtype=int32, numpy=509>

In [63]:
def f1(x, y):
    return tf.reduce_mean(input_tensor=tf.multiply(x ** 3, 6) + y**3)
func = tf.function(f1)
x = tf.constant([3., -4.])
y = tf.constant([1., 4.])
# f1 and f2 return the same value, but f2 executes as a TensorFlow graph
assert f1(x,y).numpy() == func(x,y).numpy()
#The assert passes, so there is no output

## Calculate the gradient

### GradientTape

>Another difference from numpy is that it can automatically track the gradient of any variable.

>Open one GradientTape and `tape.watch()` track variables through

In [67]:
a = tf.random.normal(shape=(2,2))
b = tf.random.normal(shape=(2,2))
with tf.GradientTape() as tape:
    tape.watch(a)
    c = tf.sqrt(tf.square(a)+tf.square(b))
    print(c)
    dc_da = tape.gradient(c,a)
    print(dc_da)

tf.Tensor(
[[1.6877743  0.97699326]
 [0.8071818  0.70809734]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[-0.16842037 -0.5020665 ]
 [ 0.9471321   0.589817  ]], shape=(2, 2), dtype=float32)


In [68]:
a

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[-0.2842556 , -0.4905156 ],
       [ 0.7645078 ,  0.41764784]], dtype=float32)>

In [69]:
b

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[ 1.6636649 ,  0.84493214],
       [ 0.25897932, -0.57181484]], dtype=float32)>

>For all variables, the calculation is tracked by default and used to find the gradient, so do not `usetape.watch()`

In [70]:
a = tf.Variable(a)
with tf.GradientTape() as tape:
    c = tf.sqrt(tf.square(a)+tf.square(b))
    dc_da = tape.gradient(c,a)
    print(dc_da)

tf.Tensor(
[[-0.16842037 -0.5020665 ]
 [ 0.9471321   0.589817  ]], shape=(2, 2), dtype=float32)


> You can GradientTapefind higher-order derivatives by opening a few more:

In [ ]:
with tf.GradientTape() as outer_tape:
    with tf.GradientTape() as tape:
        c = tf.sqrt(tf.square(a)+tf.square(b))
        dc_da = tape.gradient(c,a)
    d2c_d2a = outer_tape.gradient(dc_da,a)
    print(d2c_d2a)

tf.Tensor(
[[0.4264985  0.6867533 ]
 [0.44663432 0.5015956 ]], shape=(2, 2), dtype=float32)


# Keras, a High-Level API for TensorFlow 2

![title](img/keras.png)

## The Keras Sequential model

`To build a Keras Sequential model, you add layers to it in the same order that you want the computations to be undertaken by the network.`

`After you have built your model, you compile it; this optimizes the computations that are to be undertaken, and is where you allocate the optimizer and the loss function you want your model to use.`

`The next stage is to fit the model to the data. This is commonly known as training the model, and is where all the computations take place. It is possible to present the data to the model either in batches, or all at once.`

`Next, you evaluate your model to establish its accuracy, loss, and other metrics. Finally, having trained your model, you can use it to make predictions on new data. So, the workflow is: build, compile, fit, evaluate, make predictions.`

`There are two ways to create a Sequential model. Let's take a look at each of them.`

![title](img/kerasmodelling.png)

### Using Sequential model

`Firstly, you can pass a list of layer instances to the constructor, as in the following example.For now, we will just explain enough to allow you to understand what is happening here.`

`Acquire the data. MNIST is a dataset of hand-drawn numerals, each on a 28 x 28 pixel grid. Every individual data point is an unsigned 8-bit integer (uint8), as are the labels:`

#### Loading the datset

In [71]:
mnist_data = tf.keras.datasets.mnist
(train_x,train_y), (test_x, test_y) = mnist_data.load_data()

11490434/11490434 [==============================] - 2s 0us/step


#### Definning the variables

In [72]:
epochs=10
batch_size = 32

In [73]:
# normalize all the data points and cast the labels to int64
train_x, test_x = tf.cast(train_x/255.0, tf.float32), tf.cast(test_x/255.0, tf.float32)
train_y, test_y = tf.cast(train_y,tf.int64),tf.cast(test_y,tf.int64)

#### Building the Architecture 

In [74]:
mnistmodel1 = tf.keras.models.Sequential([
tf.keras.layers.Flatten(),
tf.keras.layers.Dense(512,activation=tf.nn.relu),
tf.keras.layers.Dropout(0.2),
tf.keras.layers.Dense(10,activation=tf.nn.softmax)
])


#### Compiling the model

In [75]:
optimiser = tf.keras.optimizers.Adam()
mnistmodel1.compile (optimizer= optimiser, loss='sparse_categorical_crossentropy', metrics = ['accuracy'])

#### Fitting the model

In [76]:
mnistmodel1.fit(train_x, train_y, batch_size=32, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 9s 2ms/step - loss: 0.2178 - accuracy: 0.9355
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0952 - accuracy: 0.9713
Epoch 3/5
1875/1875 [==============================] - 5s 2ms/step - loss: 0.0681 - accuracy: 0.9791
Epoch 4/5
1875/1875 [==============================] - 5s 2ms/step - loss: 0.0529 - accuracy: 0.9827
Epoch 5/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0448 - accuracy: 0.9854


#### Evaluate the mnistmodel1

In [77]:
mnistmodel1.evaluate(test_x, test_y)

313/313 [==============================] - 1s 2ms/step - loss: 0.0646 - accuracy: 0.9815


[0.06456564366817474, 0.9815000295639038]

>This represents a loss of 0.06 and an accuracy of 0.9815 on the test data. 

>An accuracy of 0.98 means that out of 100 test data points, 98 were, on average, correctly identified by the model.

The second way to create a Sequential model
The alternative to passing a list of layers to the Sequential model's constructor is to use the add method, as follows, for the same architecture:


#### Building the Architecture & Compiling

In [78]:
mnistmodel2 = tf.keras.models.Sequential();
mnistmodel2.add(tf.keras.layers.Flatten())
mnistmodel2.add(tf.keras.layers.Dense(512, activation='relu'))
mnistmodel2.add(tf.keras.layers.Dropout(0.2))
mnistmodel2.add(tf.keras.layers.Dense(10,activation=tf.nn.softmax))
mnistmodel2.compile (optimizer= tf.keras.optimizers.Adam(), loss='sparse_categorical_crossentropy',metrics = ['accuracy'])

#### Fitting the mnistmodel2

In [79]:
mnistmodel2.fit(train_x, train_y, batch_size=64, epochs=5)

Epoch 1/5
938/938 [==============================] - 4s 3ms/step - loss: 0.2442 - accuracy: 0.9294
Epoch 2/5
938/938 [==============================] - 4s 4ms/step - loss: 0.1047 - accuracy: 0.9679
Epoch 3/5
938/938 [==============================] - 3s 3ms/step - loss: 0.0730 - accuracy: 0.9773
Epoch 4/5
938/938 [==============================] - 3s 3ms/step - loss: 0.0545 - accuracy: 0.9830
Epoch 5/5
938/938 [==============================] - 3s 3ms/step - loss: 0.0437 - accuracy: 0.9861


#### Evaluate the mnistmodel2

In [80]:
mnistmodel2.evaluate(test_x, test_y)

313/313 [==============================] - 1s 2ms/step - loss: 0.0623 - accuracy: 0.9799


[0.0622500441968441, 0.9799000024795532]

![title](img/kerasmodelling2.png)

### Keras functional API

`The functional API lets you build much more complex architectures than the simple linear stack of Sequential models we have seen previously. It also supports more advanced models. These models include multi-input and multi-output models, models with shared layers, and models with residual connections.`

In [81]:
import tensorflow as tf
mnist = tf.keras.datasets.mnist
(train_x,train_y), (test_x, test_y) = mnist.load_data()
train_x, test_x = train_x/255.0, test_x/255.0
epochs=10


#### Building the Architecture

In [82]:
inputs = tf.keras.Input(shape=(28,28)) # Returns a 'placeholder' tensor
x = tf.keras.layers.Flatten()(inputs)
x = tf.keras.layers.Dense(512, activation='relu',name='d1')(x)
x = tf.keras.layers.Dropout(0.2)(x)
predictions = tf.keras.layers.Dense(10,activation=tf.nn.softmax, name='d2')(x)
mnistmodel3 = tf.keras.Model(inputs=inputs, outputs=predictions)

#### Compile & Fit

In [83]:
optimiser = tf.keras.optimizers.Adam()
mnistmodel3.compile (optimizer= optimiser, loss='sparse_categorical_crossentropy', metrics = ['accuracy'])
mnistmodel3.fit(train_x, train_y, batch_size=32, epochs=epochs)

Epoch 1/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2203 - accuracy: 0.9344
Epoch 2/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0981 - accuracy: 0.9699
Epoch 3/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0684 - accuracy: 0.9786
Epoch 4/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0546 - accuracy: 0.9822
Epoch 5/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0432 - accuracy: 0.9856
Epoch 6/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0368 - accuracy: 0.9879
Epoch 7/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0309 - accuracy: 0.9898
Epoch 8/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0294 - accuracy: 0.9903
Epoch 9/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0241 - accuracy: 0.9924
Epoch 10/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.023

#### Evaluate the mnistmodel3

In [84]:
mnistmodel3.evaluate(test_x, test_y)

313/313 [==============================] - 1s 2ms/step - loss: 0.0823 - accuracy: 0.9791


[0.08232200890779495, 0.9790999889373779]

### Subclassing the Keras Model class

In [85]:
import tensorflow as tf

#### Building the subclass architecture

In [86]:
class MNISTModel(tf.keras.Model):
    def __init__(self, num_classes=10):
        super(MNISTModel, self).__init__()
        # Define your layers here.
        inputs = tf.keras.Input(shape=(28,28)) # Returns a placeholder tensor
        self.x0 = tf.keras.layers.Flatten()
        self.x1 = tf.keras.layers.Dense(512, activation='relu',name='d1')
        self.x2 = tf.keras.layers.Dropout(0.2)
        self.predictions = tf.keras.layers.Dense(10,activation=tf.nn.softmax, name='d2')


    def call(self, inputs):
    # This is where to define your forward pass
    # using the layers previously defined in `__init__`
        x = self.x0(inputs)
        x = self.x1(x)
        x = self.x2(x)
        return self.predictions(x)

In [87]:
mnistmodel4 = MNISTModel()

#### Compile & Fit

In [91]:
import numpy as np

batch_size = 32
steps_per_epoch = len(train_x)//batch_size
print(steps_per_epoch)
mnistmodel4.compile (optimizer= tf.keras.optimizers.Adam(), loss='sparse_categorical_crossentropy',metrics = ['accuracy'])
mnistmodel4.fit(train_x, train_y, batch_size=batch_size, epochs=epochs)

1875
Epoch 1/10
1875/1875 [==============================] - 7s 3ms/step - loss: 0.2182 - accuracy: 0.9344
Epoch 2/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0954 - accuracy: 0.9713
Epoch 3/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0696 - accuracy: 0.9779
Epoch 4/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0530 - accuracy: 0.9831
Epoch 5/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0437 - accuracy: 0.9857
Epoch 6/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0361 - accuracy: 0.9886
Epoch 7/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0304 - accuracy: 0.9895
Epoch 8/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0289 - accuracy: 0.9902
Epoch 9/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0260 - accuracy: 0.9912
Epoch 10/10
1875/1875 [==============================] - 5s 3ms/step - loss: 

#### Evaluate the mnistmodel4

In [92]:
mnistmodel4.evaluate(test_x, test_y)

313/313 [==============================] - 1s 2ms/step - loss: 0.0682 - accuracy: 0.9831


[0.06816127896308899, 0.9830999970436096]